index.bs

```diff
@@ -2028,9 +2039,235 @@ BrowserResult = (
+    <dd>If not [=restore client window is supported=] for |window| return [=error=] with
+    [=error code=] [=unsupported operation=].
+
+    <dt>"<code>maximize</code>"
+    <dd>If not [=maximize client window is supported=] for |window| return [=error=] with
+    [=error code=] [=unsupported operation=].
+
+    <dt>"<code>minimize</code>"
+    <dd>If not [=minimize client window is supported=] for |window| return [=error=] with
+    [=error code=] [=unsupported operation=].
+   </dl>
+
+1. Let |documents| be an empty [=/list=].
+
+1. For each [=/top-level traversable=] |traversable|:
+
+  1. If |traversable|'s associated [=client window=] is not |window| then
+     continue.
+
+  1. Let |document| be |traversable|'s [=active document=].
+
+  1. Append |document| to |documents|.
+
+1. If |documents| is [=list/empty=] return [=error=] with [=error code=]
+   [=no such client window=].
+
+1. If |current state| is "<code>fullscreen</code>":
+
+  1. For each |document| in |documents|:
+
+    1. [=Fully exit fullscreen=] with |document|.
+
+       Note: This is a no-op for documents in window that are not fullscreen.
+
+1. Switch on the value of |state|:
+
+   <dl>
+    <dt>"<code>fullscreen</code>"
+    <dd>
+      1. For each |document| in |documents|:
+
+        1. If |document|'s [=visibility state=] is "<code>visible</code>",
+           [=fullscreen an element=] with |document|'s [=document element=].
+
+        1. [=Break=].
+
+    <dt>"<code>normal</code>"
+    <dd>1. [=Restore the client window=] |window|.
+
+    <dt>"<code>maximize</code>
+    <dd>1. [=Maximize the client window=] |window|.
+
+    <dt>"<code>minimize</code>
+    <dd>1. [=Minimize the client window=] |window|.
+
+1. Return [=success=] with data null.
+
+</div>
 
 ### Types ### {#module-browser-types}
 
+#### The browser.ClientWindow Type #### {#type-browser-ClientWindow}
+
+<pre class="cddl remote-cddl local-cddl">
+browser.ClientWindow = text;
+</pre>
+
+The <code>browser.ClientWindow</code> uniquely identifies a [=client window=].
+
+#### The browser.ClientWindowInfo Type #### {#type-browser-ClientWindowInfo}
+
+<pre class="cddl remote-cddl local-cddl">
+browser.ClientWindowInfo = {
+  clientWindow: browser.ClientWindow,
+  hasFocus: bool,
+  height: js-uint,
+  state: "fullscreen" / "maximized" / "minimized" / "normal",
+  width: js-uint,
+  x: js-int,
+  y: js-int,
+}
+</pre>
+
+The <code>browser.ClientWindowInfo</code> type represents properties of a
+[=client window=].
+
+<div algorithm>
+To <dfn>get the client window info</dfn> given |client window|:
+
+1. Let |client window id| be the [=client window id=] for |client window|.
+
+1. Let |state| be [=get the client window state=] with |client window|.
+
+1. Let |has system focus| be false.
+
+1. For each [=/top-level traversable=] |traversable|:
+
+  1. If |traversable|'s [=client window=] is not |client window| then continue.
+
+  1. If |traversable| has [=system focus=], set |has system focus| to true and break.
```

Comment:
Should we actually get this information from [`taversable's document's hasFocus()`](https://html.spec.whatwg.org/#dom-document-hasfocus)?

Reply:
Actually I think that technically neither is correct and we should define our own thing (see the example about the address bar having focus; in that case the client window has focus, but no traversable does).